arch/sim/src/sim/up_modtext.c

```diff
@@ -0,0 +1,104 @@
+/****************************************************************************
+ * arch/sim/src/sim/up_modtext.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/* Note: this module assumes that size_t is compatible between
+ * nuttx and host.
+ */
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/mman.h>
+
+#include <assert.h>
+#include <stddef.h>
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct module_text_header
+{
+  size_t size;
+};
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_module_text_init
+ *
+ * Description:
+ *   Initialize the module text allocator
+ *
+ ****************************************************************************/
+
+void up_module_text_init()
+{
+}
+
+/****************************************************************************
+ * Name: up_module_text_alloc
+ *
+ * Description:
+ *   Allocate memory for module text.
+ *
+ ****************************************************************************/
+
+void *up_module_text_alloc(size_t size)
+{
+  struct module_text_header *hdr;
+
+  hdr = mmap(NULL, sizeof(*hdr) + size, PROT_READ | PROT_WRITE | PROT_EXEC,
```

Comment:
Why we can't directly use the main heap?

Reply:
from the commit message:

> On modern environments, sim_heap on bss is not executable.
> Enable only on macOS for now because I haven't investigated others.

Reply:
> from the commit message:
> 
> > On modern environments, sim_heap on bss is not executable.
> > Enable only on macOS for now because I haven't investigated others.

we can change sim_heap to mmap/mumap directly, this change is more simpler.

Reply:
maybe. but in general it's a good thing to have the heap non executable.

Reply:
> maybe. but in general it's a good thing to have the heap non executable.

But, the new change make the code much complex. The most usage for sim is for developing and the security impact of executable heap is small.

Reply:
ok, it makes sense.

Reply:
done.